esp/src/src-react/components/Metrics.tsx

```diff
@@ -0,0 +1,271 @@
+                updateMetricGraph(rows);
+            }
+        })
+    );
+
+    React.useEffect(() => {
+        scopesTable
+            .columns(["##", nlsHPCC.Type, nlsHPCC.Scope, ...options.properties])
+            .data(metrics.filter(row => {
+                return (timelineFilter === "" || row.name?.indexOf(timelineFilter) === 0) &&
+                    (options.scopeTypes.indexOf(row.type) >= 0);
+            }).map((row, idx) => {
+                row.__hpcc_id = row.id;
+                return [idx, row.type, row.name, ...options.properties.map(p => row[p] !== undefined ? row[p] : ""), row];
+            }))
+            .lazyRender()
+            ;
+    }, [hasFilter, metrics, scopesTable, timelineFilter, filter, options.properties, options.scopeTypes]);
+
+    //  Graph  ---
+    const metricGraph = useConst(() => new MetricGraph());
+    const metricGraphWidget = useConst(() => new MetricGraphWidget()
+        .id("metricGraph")
+        .zoomToFitLimit(1)
+        .on("selectionChanged", () => {
+            const items = metricGraphWidget.selection().map(id => {
+                return metricGraph.item(id);
+            });
+            // const item = graph.item(row.id);
+            updatePropsTable(items);
+            updatePropsTable2(items);
+            const tableItems = scopesTable.data().filter(tableRow => items.indexOf(tableRow[tableRow.length - 1]) >= 0);
+            scopesTable.selection(tableItems);
+        })
+    );
+
+    React.useEffect(() => {
+        metricGraph.load(metrics);
+    }, [metrics, metricGraph]);
+
+    const updateMetricGraph = React.useCallback((selection: IScope[]) => {
+        if (selection.length) {
+            metricGraphWidget
+                .dot(metricGraph.graphTpl(selection, options))
+                .resize()
+                .render(() => {
+                    metricGraphWidget.selection(selection.map(s => s.id));
+                })
+                ;
+        }
+    }, [metricGraph, metricGraphWidget, options]);
+
+    //  Props Table  ---
+    const propsTable = useConst(() => new Table()
+        .id("propsTable")
+        .columns([nlsHPCC.Property, nlsHPCC.Value])
+        .columnWidth("none")
+    );
+
+    const updatePropsTable = items => {
+        const props = [];
+        items.forEach((item, idx) => {
+            for (const key in item) {
+                if (key.indexOf("__") !== 0) {
+                    props.push([key, item[key]]);
+                }
+            }
+            if (idx < items.length - 1) {
+                props.push(["------------------------------", "------------------------------"]);
+            }
+        });
+        propsTable
+            ?.data(props)
+            ?.lazyRender()
+            ;
+    };
+
+    const propsTable2 = useConst(() => new Table()
+        .id("propsTable2")
+        .columns([nlsHPCC.Property, nlsHPCC.Value])
+        .columnWidth("auto")
+    );
+
+    const updatePropsTable2 = items => {
```

Comment:
Same. should this also be a hook, useCallback?

Reply:
Agree.

Reply:
(and for the cases here...)